### PR TITLE
Add spaces to template literals when using parenSpacing

### LINKED ISSUE #19: Inconsistent Whitespace Inside Template Strings After the Upgrade to 2.0.4
After the upgrade to WP Prettier 2.0.4, I’m getting inconsistent whitespace padding inside template strings:

```diff
--- a/src/blocks/sections/hero/html/heading.js
+++ b/src/blocks/sections/hero/html/heading.js
@@ -16,7 +16,7 @@ export default ( { className, children, isImagePresent } ) => {
                'glp-text-brand',
                'glp-text-32',
                'glp-text-40@S',
-               `glp-text-${ isImagePresent ? 56 : 64 }@M`,
+               `glp-text-${ isImagePresent ? 56 : 64}@M`,
                'glp-color-primary',
                className,
        ] );
```

The same happens to all other template strings in the project.

#### Input
```jsx
`glp-text-${ isImagePresent ? 56 : 64 }@M`
```

#### Output
```jsx
`glp-text-${ isImagePresent ? 56 : 64}@M`
```

#### Expected behavior
```jsx
// The previous behavior, consistent with WP standards:
`glp-text-${ isImagePresent ? 56 : 64 }@M`

// This one would also be fine with me:
`glp-text-${isImagePresent ? 56 : 64}@M`
```

### PR DESCRIPTION
**Pull request by scinos**
Changes:
* It adds an extra space to template literals when using `parenSpace`
* Adds lots of tests for parenSpacing

Fixes #19

### CODE REVIEW

File: tests/paren_space/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,802 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`arrays.js 1`] = `
+====================================options=====================================
+parenSpacing: false
+parsers: ["babel"]
+printWidth: 80
+                                                                                | printWidth
+=====================================input======================================
```
Comment (scinos) [inline tests/paren_space/__snapshots__/jsfmt.spec.js.snap:9]: How to read this file: This is a set of blocks `input` and `output`. When given the `input` entry, prettier will transform it to the next `output` block.

**Comment by scinos** [comment]
To generate the tests, I copy/pasted a few existing tests from `prettier` (so they covers most of what prettier js covers), specially those related to parentheses.

Then I enabled `parenSpacing: true` and re-generated the snapshots.

For reference, these are the differences between `parenSpacing: true` and `parenSpacing: false` (i.e. the difference between wp-prettier and prettier, so to speak): 

[spacing.diff.txt](https://github.com/Automattic/wp-prettier/files/4506320/spacing.diff.txt) (download it and open it with your favourite diff viewer)

File: README.md [hunk 2]
```diff
@@ -65,9 +65,11 @@
 This is a fork of Prettier that adds a new command line option `--paren-spacing` which inserts many extra spaces inside parentheses, the way how projects in the WordPress ecosystem (Calypso, Gutenberg, etc.) like to format their code.
 
 In order to install a version based on a particular upstream version (like 1.x.x), run
+
 ```sh
 npm i --save-dev "git+https://github.com/Automattic/wp-prettier.git#wp-prettier-1.x.x"
```
Comment (sirreal) [inline README.md]: I know it's not new, but this is stale. It should now be something like `"prettier": "npm:wp-prettier@latest"`

File: src/language-js/printer-estree.js [hunk 3]
```diff
@@ -2560,7 +2560,11 @@ function printPathNoParens(path, options, print, args) {
               isBinaryish(n.expressions[i])
             ) {
               printed = concat([
-                indent(concat([parenLine, printed])),
+                // If we are using parenSpace, parenLine already contains a space, so this
+                // will be equivalent to `" " + printed + " "`
+                // If we are not using parenSpace, then parenSpace is an empty string, so
+                // this will be equivalent to `"" + printed + ""`
+                indent(concat([parenLine, printed, parenSpace])),
                 softline,
```
Comment (sirreal) [inline src/language-js/printer-estree.js]: Let's go with the approach (that I mangled on the 2.0.4 branch) from `1.19.1`:

https://github.com/Automattic/wp-prettier/blob/da6be4794765d435c2f42a3218545cca679a0454/src/language-js/printer-estree.js#L2526-L2529

**Comment by sirreal** [comment]
Here's the diff in an easy to review format:

<details>
<summary>Expand to view the diff</summary>

```diff
diff --git b/tests/paren_space/__snapshots__/jsfmt.spec.js.snap a/tests/paren_space/__snapshots__/jsfmt.spec.js.snap
index ee642b8b..7982c666 100644
--- b/tests/paren_space/__snapshots__/jsfmt.spec.js.snap
+++ a/tests/paren_space/__snapshots__/jsfmt.spec.js.snap
@@ -18,15 +18,15 @@ printWidth: 80
 [...a, ...b];
 
 =====================================output=====================================
-[[]];
-[[], []];
-[[], [], []];
-[[], [0], []];
-[[], [0], [0]];
-[[], [0, 1], [0]];
-[[], [0, 1], [0, 1]];
-[...a, ...b];
-[...a, ...b];
+[ [] ];
+[ [], [] ];
+[ [], [], [] ];
+[ [], [ 0 ], [] ];
+[ [], [ 0 ], [ 0 ] ];
+[ [], [ 0, 1 ], [ 0 ] ];
+[ [], [ 0, 1 ], [ 0, 1 ] ];
+[ ...a, ...b ];
+[ ...a, ...b ];
 
 ================================================================================
 `;
@@ -78,44 +78,44 @@ a => ({}:: b()\`\`[''].c++ && 0 ? 0 : 0);
 a:: (b => c);
 
 =====================================output=====================================
-(a || b)::c;
+( a || b )::c;
 a || b::c;
 ::obj.prop;
-(void 0)::func();
-(+0)::is(-0);
+( void 0 )::func();
+( +0 )::is( -0 );
 a::b.c;
-a::(b.c());
+a::( b.c() );
 a::b.c();
-a::(b.c()());
-a::(b.c()());
-a::(b.c())();
-a::(b.c().d);
-a::(c().d.e);
-a::(b());
-a::(b::c());
-a::(b()::c);
-a::(b().c::d);
-a::(b.c::d);
-a::(b::c.d);
-a::(b.c::d::e);
-a::(b::c::d);
-a::(b::c::d.e);
-a::(b::c::d).e;
-a::(void 0);
-a::(b.c()::d.e);
-a::(b.c::d.e);
-a::(b.c::d.e)::f.g;
+a::( b.c()() );
+a::( b.c()() );
+a::( b.c() )();
+a::( b.c().d );
+a::( c().d.e );
+a::( b() );
+a::( b::c() );
+a::( b()::c );
+a::( b().c::d );
+a::( b.c::d );
+a::( b::c.d );
+a::( b.c::d::e );
+a::( b::c::d );
+a::( b::c::d.e );
+a::( b::c::d ).e;
+a::( void 0 );
+a::( b.c()::d.e );
+a::( b.c::d.e );
+a::( b.c::d.e )::f.g;
 b.c::d.e;
-(b.c::d).e;
-(b::c::d).e;
-new (a::b)();
-new f(a::b);
-f[a::b];
-f[a::b()];
+( b.c::d ).e;
+( b::c::d ).e;
+new ( a::b )();
+new f( a::b );
+f[ a::b ];
+f[ a::b() ];
 
-(a) => ({}::b()\`\`[""].c++ && 0 ? 0 : 0);
-((a) => b)::c;
-a::((b) => c);
+( a ) => ( {}::b()\`\`[ "" ].c++ && 0 ? 0 : 0 );
+( ( a ) => b )::c;
+a::( ( b ) => c );
 
 ================================================================================
 `;
@@ -144,13 +144,13 @@ class X {
 
 =====================================output=====================================
 class c {
-  ["i"]() {}
+  [ "i" ]() {}
 }
 
-(class {} + 1);
-(class a {} + 1);
-(class extends b {} + 1);
-(class a extends b {} + 1);
+( class {} + 1 );
+( class a {} + 1 );
+( class extends b {} + 1 );
+( class a extends b {} + 1 );
 
 class X {
   async *b() {
@@ -205,17 +205,17 @@ if (
 }
 
 const radioSelectedAttr =
-  (isAnyValueSelected &&
-    node.getAttribute(radioAttr.toLowerCase()) === radioValue) ||
-  (!isAnyValueSelected && values[a].default === true) ||
+  ( isAnyValueSelected &&
+    node.getAttribute( radioAttr.toLowerCase() ) === radioValue ) ||
+  ( ! isAnyValueSelected && values[ a ].default === true ) ||
   a === 0;
 
 function f() {
-  if (position) return { name: pair };
+  if ( position ) return { name: pair };
   else
     return {
-      name: pair.substring(0, position),
-      value: pair.substring(position + 1),
+      name: pair.substring( 0, position ),
+      value: pair.substring( position + 1 ),
     };
 }
 
@@ -241,8 +241,8 @@ a = ([s = 1,]) => 1
 const { children, ...props } = this.props
 
 =====================================output=====================================
-const [one, two = null, three = null] = arr;
-a = ([s = 1]) => 1;
+const [ one, two = null, three = null ] = arr;
+a = ( [ s = 1 ] ) => 1;
 const { children, ...props } = this.props;
 
 ================================================================================
@@ -272,19 +272,19 @@ new (factory())(factory())
 object.foo().bar().baz();
 
 =====================================output=====================================
-(this.x++).toString();
+( this.x++ ).toString();
 
-new (r++)();
+new ( r++ )();
 
-(x++)();
+( x++ )();
 
-const uuid = String(this._uuidCounter++);
+const uuid = String( this._uuidCounter++ );
 
-!(
+! (
   { a: 1, b: 2 } // foo
 );
 
-new (factory())(factory());
+new ( factory() )( factory() );
 
 object.foo().bar().baz();
 
@@ -320,24 +320,24 @@ new function () { };
 
 =====================================output=====================================
 function* f() {
-  yield async (a) => a;
+  yield async ( a ) => a;
 }
 
 async function f3() {
-  a = (await 1) ? 1 : 1;
+  a = ( await 1 ) ? 1 : 1;
 }
 
-(function () {}.length);
+( function () {}.length );
 typeof function () {};
-export default (function () {})();
-(function () {})()\`\`;
-(function () {})\`\`;
-new (function () {})();
-(function () {});
+export default ( function () {} )();
+( function () {} )()\`\`;
+( function () {} )\`\`;
+new ( function () {} )();
+( function () {} );
 a = function f() {} || b;
-(function () {} && a);
+( function () {} && a );
 a + function () {};
-new (function () {})();
+new ( function () {} )();
 
 ================================================================================
 `;
@@ -362,13 +362,13 @@ export default (function f() { })
 =====================================output=====================================
 import { fitsIn, oneLine } from ".";
 
-import("module.js").then((a) => a);
+import( "module.js" ).then( ( a ) => a );
 
 export a, { b } from "./baz";
 
 export default () => {};
 
-export default (function f() {});
+export default ( function f() {} );
 
 ================================================================================
 `;
@@ -415,16 +415,16 @@ do {} while (
   )
 );
 
-for (var i = 0, len = arr.length; i < len; i++) {}
+for ( var i = 0, len = arr.length; i < len; i++ ) {}
 
-(async () => {
-  for await (num of asyncIterable) {
-    console.log(num);
+( async () => {
+  for await ( num of asyncIterable ) {
+    console.log( num );
   }
-  for await (const x of delegate_yield()) {
+  for await ( const x of delegate_yield() ) {
     x;
   }
-})();
+} )();
 
 ================================================================================
 `;
@@ -444,7 +444,7 @@ foo ?? (bar ?? baz);
 (foo ?? bar) ?? baz;
 
 =====================================output=====================================
-const x = (foo, bar = foo ?? bar) => {};
+const x = ( foo, bar = foo ?? bar ) => {};
 
 foo ? bar ?? foo : baz;
 
@@ -472,15 +472,15 @@ a = () => ({}).x;
 (({} = 0), 1);
 
 =====================================output=====================================
-() => ({}\`\`);
-({}\`\`);
-a = () => ({}.x);
-({} && a, b);
-({}::b, 0);
-({}::b()\`\`[""].c++ && 0 ? 0 : 0, 0);
-({}(), 0);
-({} = 0);
-({} = 0), 1;
+() => ( {}\`\` );
+( {}\`\` );
+a = () => ( {}.x );
+( {} && a, b );
+( {}::b, 0 );
+( {}::b()\`\`[ "" ].c++ && 0 ? 0 : 0, 0 );
+( {}(), 0 );
+( {} = 0 );
+( {} = 0 ), 1;
 
 ================================================================================
 `;
@@ -526,31 +526,31 @@ printWidth: 80
   (a = b ? c : function () { return 0; });
 
 =====================================output=====================================
-(a = b
+( a = b
   ? c
   : function () {
       return 0;
-    }),
-  (a = b
+    } ),
+  ( a = b
     ? c
     : function () {
         return 0;
-      }),
-  (a = b
+      } ),
+  ( a = b
     ? c
     : function () {
         return 0;
-      }),
-  (a = b
+      } ),
+  ( a = b
     ? c
     : function () {
         return 0;
-      }),
-  (a = b
+      } ),
+  ( a = b
     ? c
     : function () {
         return 0;
-      });
+      } );
 
 ================================================================================
 `;
@@ -571,14 +571,14 @@ async () => ({ ...(await foo) });
 declare class C { f(...r): void; }
 
 =====================================output=====================================
-const foo = { ...(a || b) };
-const foo2 = { ...(a || b) };
-const foo3 = { ...(a ? b : c) };
+const foo = { ...( a || b ) };
+const foo2 = { ...( a || b ) };
+const foo3 = { ...( a ? b : c ) };
 
-async () => ({ ...(await foo) });
+async () => ( { ...( await foo ) } );
 
 declare class C {
-  f(...r): void;
+  f( ...r ): void;
 }
 
 ================================================================================
@@ -606,18 +606,18 @@ switch (a) {
 }
 
 =====================================output=====================================
-switch (a) {
+switch ( a ) {
   case 3:
-    alert("3");
+    alert( "3" );
     break;
   case 4:
-    alert("4");
+    alert( "4" );
     break;
   case 5:
-    alert("5");
+    alert( "5" );
     break;
   default:
-    alert("default");
+    alert( "default" );
 }
 
 ================================================================================
@@ -667,9 +667,9 @@ return (<div css={\`
 \`}/>)
 
 =====================================output=====================================
-\`\${1 ?? 2}\`;
+\`\${ 1 ?? 2 }\`;
 
-\`glp-text-\${isImagePresent ? 56 : 64}@M\`;
+\`glp-text-\${ isImagePresent ? 56 : 64 }@M\`;
 
 const headerResolve = css.resolve\`
   .top-bar {
@@ -695,7 +695,7 @@ const headerResolve = css.resolve\`
 
 return (
   <div
-    css={\`
+    css={ \`
       color: blue;
       font-size: 17 px;
 
@@ -706,7 +706,7 @@ return (
       & .some-class {
         font-size: 20px;
       }
-    \`}
+    \` }
   />
 );
 
@@ -723,7 +723,7 @@ printWidth: 80
 a => a ? () => { a } : () => { a }
 
 =====================================output=====================================
-(a) =>
+( a ) =>
   a
     ? () => {
         a;
@@ -755,9 +755,9 @@ function f() {
   throw foo.bar();
 }
 
-lint(ast, {
-  with: () => throw new Error("avoid using 'with' statements."),
-});
+lint( ast, {
+  with: () => throw new Error( "avoid using 'with' statements." ),
+} );
 
 ================================================================================
 `;
@@ -776,7 +776,7 @@ finally { await baz() }
 =====================================output=====================================
 try {
   foo();
-} catch (e) {
+} catch ( e ) {
   bar();
 } finally {
   await baz();
@@ -795,7 +795,7 @@ printWidth: 80
 with (0) { };
 
 =====================================output=====================================
-with (0) {
+with ( 0 ) {
 }
 
 ================================================================================

```

</details>

**Comment by sirreal** [comment]
The paren spacing snapshot diff looks about right to me. Let's apply this change (https://github.com/Automattic/wp-prettier/pull/20/files#r411954955) verify it's working correctly, then merge.

File: src/language-js/printer-estree.js (hunk 3 above)
Comment (scinos) [reply src/language-js/printer-estree.js]: Done

File: README.md (hunk 2 above)
Comment (scinos) [reply README.md]: Updated

**Comment by scinos** [comment]
I can't figure out why the tests are failing. What I know is they are not related to this particular change.

So I propose merge this as-is to fix #19, release it as `wp-prettier@2.0.5-alpha-1`, and investigate the broken test later.

@sirreal @sgomes @griffbrad @jsnajdr  opinions?

Note about the numbering scheme:
- 2.0.4 is used, and anything like `2.0.4-xxx` will have a lower precedence (so people won't get it if they already have 2.0.4)
- We could use `2.0.5-alpha-1`, `2.0.5-alpha-2`... for all the patches we apply on top of `2.0.4`.
- When prettier releases `2.0.5`, we merge it an release it as `2.0.5-beta-1`
- If we need to fix things, we can do `2.0.5-beta-2`...
- When we are confident that our `2.0.5-beta-xxx` series is stable, we can release it as just `2.0.5`.

Ugly, but I can't figure out a less-ugly option. Open to suggestions

**Review by lynnbgriffith — APPROVED** [review]
👍🏻 Approach to the current fix and for versioning make sense to me.

**Comment by sirreal** [comment]
I can't reproduce the CI failures locally either 😕 

I did notice that the checkout is a _merge_, it merges this branch into `wp-prettier-2.0.4` before running the tests. I've done that locally and run `yarn test --clearCache; yarn test --ci`. Everything still passes.

**Review by sirreal — APPROVED** [review]

**Pull request merged**